### ISSUE TITLE
JavaScript heap out of memory after start

### ISSUE THREAD

**Issue by dkaparis** [unwriter/bitd#4]
On startup, after going through "getBlockHashes" and then "getTransactionHeaders" up to the last block index, I get the following:

```
<--- Last few GCs --->

[25975:0x55c2e9434980]   235170 ms: Mark-sweep 1392.5 (1420.5) -> 1392.5 (1420.5) MB, 1170.8 / 0.0 ms  (average mu = 0.123, current mu = 0.008) allocation failure scavenge might not succeed
[25975:0x55c2e9434980]   236404 ms: Mark-sweep 1393.4 (1420.5) -> 1393.4 (1421.0) MB, 1230.5 / 0.0 ms  (average mu = 0.065, current mu = 0.002) allocation failure scavenge might not succeed


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x12e56ffdc01d]
Security context: 0x29df7e11e549 <JSObject>
    1: /* anonymous */ [0xaba2982291] [/.../bitd/bit.js:~191] [pc=0x12e5702b4fd7](this=0x38fd1e31aa09 <JSGlobal Object>,transactionHeader=0x32a71c38d5b9 <Object map = 0x2aedaddbef29>)
    2: arguments adaptor frame: 3->1
    3: StubFrame [pc: 0x12e56ffa6734]
    4: map [0x29df7e106cc1](this=0x19d7aad022d1 <JSArray[7710467]>,0x00aba2982291 ...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x55c2e6ffd101 node::Abort() [node]
 2: 0x55c2e6ffd14f  [node]
 3: 0x55c2e71bed72 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0x55c2e71befcb v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0x55c2e7553313  [node]
 6: 0x55c2e7553454  [node]
 7: 0x55c2e75630d3 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
 8: 0x55c2e7563a59 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 9: 0x55c2e7565b6d v8::internal::Heap::AllocateRawWithLigthRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
10: 0x55c2e7565bc2 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
11: 0x55c2e7533d65 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [node]
12: 0x55c2e77a63ff v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [node]
13: 0x12e56ffdc01d 
Aborted (core dumped)
```
I tried running node with --max-old-space-size=8000 option and I got this:

```
<--- Last few GCs --->

[26048:0x557923a42980]   234875 ms: Mark-sweep 1403.3 (1431.0) -> 1403.4 (1434.5) MB, 1505.2 / 0.0 ms  (average mu = 0.210, current mu = 0.008) allocation failure scavenge might not succeed
[26048:0x557923a42980]   236244 ms: Mark-sweep 1407.0 (1434.5) -> 1407.2 (1442.5) MB, 1359.1 / 0.0 ms  (average mu = 0.117, current mu = 0.007) allocation failure scavenge might not succeed


<--- JS stacktrace --->
Cannot get stack trace in GC.
FATAL ERROR: MarkCompactCollector: semi-space copy, fallback in old gen Allocation failed - JavaScript heap out of memory
 1: 0x55792189a101 node::Abort() [node]
 2: 0x55792189a14f  [node]
 3: 0x557921a5bd72 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0x557921a5bfcb v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0x557921df0313  [node]
 6: 0x557921e1b478 v8::internal::EvacuateNewSpaceVisitor::Visit(v8::internal::HeapObject*, int) [node]
 7: 0x557921e1fc56 void v8::internal::LiveObjectVisitor::VisitBlackObjectsNoFail<v8::internal::EvacuateNewSpaceVisitor, v8::internal::MajorNonAtomicMarkingState>(v8::internal::MemoryChunk*, v8::internal::MajorNonAtomicMarkingState*, v8::internal::EvacuateNewSpaceVisitor*, v8::internal::LiveObjectVisitor::IterationMode) [node]
 8: 0x557921e27178 v8::internal::FullEvacuator::RawEvacuatePage(v8::internal::Page*, long*) [node]
 9: 0x557921e16ada v8::internal::Evacuator::EvacuatePage(v8::internal::Page*) [node]
10: 0x557921e16e8b v8::internal::PageEvacuationTask::RunInParallel() [node]
11: 0x557921e0fe8f v8::internal::ItemParallelJob::Task::RunInternal() [node]
12: 0x557921e10a6e v8::internal::ItemParallelJob::Run(std::shared_ptr<v8::internal::Counters>) [node]
13: 0x557921e1c4da void v8::internal::MarkCompactCollectorBase::CreateAndExecuteEvacuationTasks<v8::internal::FullEvacuator, v8::internal::MarkCompactCollector>(v8::internal::MarkCompactCollector*, v8::internal::ItemParallelJob*, v8::internal::RecordMigratedSlotVisitor*, v8::internal::MigrationObserver*, long) [node]
14: 0x557921e25a37 v8::internal::MarkCompactCollector::EvacuatePagesInParallel() [node]
15: 0x557921e25be0 v8::internal::MarkCompactCollector::Evacuate() [node]
16: 0x557921e2654b v8::internal::MarkCompactCollector::CollectGarbage() [node]
17: 0x557921dffaa1 v8::internal::Heap::MarkCompact() [node]
18: 0x557921e00113 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
19: 0x557921e00a59 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
20: 0x557921e02b6d v8::internal::Heap::AllocateRawWithLigthRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
21: 0x557921e02bc2 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
22: 0x557921dd0d65 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [node]
23: 0x5579220433ff v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [node]
24: 0xc7c1fdc01d 
Aborted (core dumped)
```

**Comment by unwriter** [comment]
TLDR: Try starting with

```
node index --max-old-space-size=2048
```

or if that doesn't work,

```
node index --max-old-space-size=4096
```

Now for the explanation on why this happened: The reason this happened was because the original bitdb was designed with reasonable block size (not 20MB like caused by today's stress test), which is why there are several places where it makes bad memory management practices.

This is a known problem which will not happen in the new version I'm working on because I rewrote the whole thing from scratch. In fact I have multiple of them running on the side concurrently and they have been doing fine during the stress test.
